File: 432. All O`one Data Structure.cpp
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <string>

using namespace std;

class AllOne {
    struct Bucket {
        int count;
        unordered_set<string> keys;
    };

    list<Bucket> bucketList;
    unordered_map<string, int> keyCount;
    unordered_map<int, list<Bucket>::iterator> countBucket;

public:
    AllOne() {}

    void inc(string key) {
        int count = keyCount[key]++;
        auto curIt = count ? countBucket[count] : bucketList.end();
        auto nextIt = curIt;

        if (countBucket.find(count + 1) == countBucket.end()) {
            if (curIt == bucketList.end()) {
                bucketList.push_front({count + 1, {key}});
                countBucket[count + 1] = bucketList.begin();
            } else {
                nextIt = bucketList.insert(nextIt, {count + 1, {key}});
                countBucket[count + 1] = nextIt;
            }
        } else {
            countBucket[count + 1]->keys.insert(key);
        }

        if (count) {
            curIt->keys.erase(key);
            if (curIt->keys.empty()) {
                countBucket.erase(count);
                bucketList.erase(curIt);
            }
        }
    }

    void dec(string key) {
        int count = keyCount[key]--;
        auto curIt = countBucket[count];

        if (count == 1) {
            keyCount.erase(key);
        } else {
            if (countBucket.find(count - 1) == countBucket.end()) {
                auto newIt = bucketList.insert(curIt, {count - 1, {key}});
                countBucket[count - 1] = newIt;
            } else {
                countBucket[count - 1]->keys.insert(key);
            }
        }

        curIt->keys.erase(key);
        if (curIt->keys.empty()) {
            countBucket.erase(count);
            bucketList.erase(curIt);
        }
    }

    string getMaxKey() {
        return bucketList.empty() ? "" : *bucketList.back().keys.begin();
    }

    string getMinKey() {
        return bucketList.empty() ? "" : *bucketList.front().keys.begin();
    }
};

File: 432. All O`one Data Structure.java
import java.util.*;

class AllOne {

    class Node {
        int count;
        Set<String> keys;
        Node prev, next;

        Node(int count) {
            this.count = count;
            this.keys = new LinkedHashSet<>(); // preserves insertion order
        }
    }

    private Map<String, Node> keyCountMap;
    private Node head, tail;

    public AllOne() {
        keyCountMap = new HashMap<>();
        head = new Node(0);
        tail = new Node(0);
        head.next = tail;
        tail.prev = head;
    }

    // Insert a new node with count next to prevNode
    private Node insertAfter(Node prevNode, int count) {
        Node newNode = new Node(count);
        newNode.prev = prevNode;
        newNode.next = prevNode.next;
        prevNode.next.prev = newNode;
        prevNode.next = newNode;
        return newNode;
    }

    // Remove a node if its key set is empty
    private void removeNode(Node node) {
        if (node != head && node != tail && node.keys.isEmpty()) {
            node.prev.next = node.next;
            node.next.prev = node.prev;
        }
    }

    public void inc(String key) {
        Node curr = keyCountMap.getOrDefault(key, head);
        Node nextNode;

        if (curr.next == tail || curr.next.count != curr.count + 1) {
            nextNode = insertAfter(curr, curr.count + 1);
        } else {
            nextNode = curr.next;
        }

        nextNode.keys.add(key);
        if (curr != head) {
            curr.keys.remove(key);
            removeNode(curr);
        }

        keyCountMap.put(key, nextNode);
    }

    public void dec(String key) {
        Node curr = keyCountMap.get(key);
        if (curr == null) return;

        if (curr.count == 1) {
            keyCountMap.remove(key);
        } else {
            Node prevNode;
            if (curr.prev == head || curr.prev.count != curr.count - 1) {
                prevNode = insertAfter(curr.prev, curr.count - 1);
            } else {
                prevNode = curr.prev;
            }

            prevNode.keys.add(key);
            keyCountMap.put(key, prevNode);
        }

        curr.keys.remove(key);
        removeNode(curr);
    }

    public String getMaxKey() {
        return tail.prev != head && !tail.prev.keys.isEmpty()
                ? tail.prev.keys.iterator().next()
                : "";
    }

    public String getMinKey() {
        return head.next != tail && !head.next.keys.isEmpty()
                ? head.next.keys.iterator().next()
                : "";
    }
}
